docs: add Miles server arguments (#517)
radixark/miles: docs/en/advanced/miles_server_args.md

| `--apply-chat-template-kwargs` | Extra arguments for the chat template processing (JSON string). | `"{}"` | Type: str | Miles Native |
| `--num-rollout` | Number of rollout steps. If not set, Miles will calculate the number of rollout steps from the dataset size. **Note:** This value will be overwritten if `--num-epoch` is also set. | `None` | Type: int | Miles Native |
| `--num-epoch` | Number of epochs for the training. If set, `num_rollout` is calculated as `(num_epoch * dataset_size) // rollout_batch_size`. **Note:** This argument takes precedence and will overwrite `--num-rollout` if both are specified. | `None` | Type: int | Miles Native |
| `--rollout-batch-size` | Number of prompts per rollout batch. The total data returned should be `rollout_batch_size` * `n_samples_per_prompt`. | Required | Type: int | Miles Native |
| `--n-samples-per-prompt` | Number of responses to generate for each prompt, e.g., the group size of GRPO. | `1` | Type: int | Miles Native |
| `--global-batch-size` | Total samples per optimizer step. Automatically calculated or **overridden** if `num_steps_per_rollout` is set. | `None` | Type: int | Megatron-LM (Reset by Miles) |

> | `--num-steps-per-rollout` | The number of training steps to perform using the data collected in a single rollout round. Setting this to `n` means the policy model will be updated `n` times using the same batch of rollout data. Miles ensures that `(rollout-batch-size * n-samples-per-prompt) = (global-batch-size * num-steps-per-rollout)`. If this value is not provided, you have to set `--global-batch-size` explicitly. If both are provided, `--num-steps-per-rollout`  will **override** the global batch size with `num_steps_per_rollout = (rollout_batch_size * n_samples_per_prompt) // num_steps_per_rollout`. | `None` | Type: int | Miles Native |

Reviewer: Could you check what will happend if multi samples are returned in the generate function (maybe the refactored one)?

| `--log-probs-max-tokens-per-gpu` | The maximum number of tokens per GPU for calculating log probs. This is used to calculate the log probs of the responses during rollout, and should be set to a larger value than `max_tokens_per_gpu` if you want better performance. | `None` | Type: int | Miles Native |
| `--balance-data` | Repartition each rollout batch so each data-parallel rank gets a similar total token count via the Karmarkar-Karp method. It may be beneficial for training speed, but changes per-rank sample grouping and adds a small CPU scheduling overhead. | `False` | bool flag (set to enable) | Miles Native |
| `--data-pad-size-multiplier` | Multiplier used to calculate the sequence padding boundary. Miles rounds sequence lengths up to a multiple of `tensor_parallel_size * data_pad_size_multiplier`. This optimization ensures that matrix dimensions are aligned with NVIDIA Tensor Core requirements, maximizing throughput and reducing VRAM fragmentation. | `128` | Type: int | Miles Native |

> | `--micro-batch-size` | Micro batch size per GPU. Ignored when `--use-dynamic-batch-size` is enabled. | `1` | Type: int | Megatron-LM (Reset by Miles) |

Reviewer: Not work for `--qkv-format=thd`

> 

Reviewer: Could you add one more parameter `--seq-length`? It's a very confusing param in Megatron but not effective in miles at all. Ref https://github.com/radixark/miles/issues/574#issuecomment-3873918902

| `--disable-grpo-std-normalization` | Disable standard deviation normalization for GRPO. From [Dr.GRPO](https://arxiv.org/pdf/2503.20783) | `False` | bool flag (set to enable) | Miles Native |
| `--disable-rewards-normalization` | Disable the default group-wise reward normalization for GRPO, GSPO, and REINFORCE++. This effectively skips the baseline subtraction step. | `False` | bool flag (set to enable) | Miles Native |
| `--use-rollout-entropy` | Enable entropy calculation when calculating the logprobs from actor and reference model. This is useful for implementing custom entropy-based loss masking. | `False` | bool flag (set to enable) | Miles Native |
| `--use-rollout-logprobs` | Use rollout logprobs for importance sampling ratios, use the logprobs from the actor model if not set. If `--get-mismatch-metrics` is set, the log probs will be recomputed by training engine, one more forward pass will be applied. | `False` | bool flag (set to enable) | Miles Native |
| `--calculate-per-token-loss` | Calculate loss on a per-token basis. | `False` | bool flag (set to enable) | Megatron-LM (Reset by Miles) |

> | `--seed` | Random seed for the training process. | `1234` | Type: int | Megatron-LM (Reset by Miles) |

Reviewer: This `--seed` also affect sglang server's `--random-seed` param.

| `--enable-mtp-training` | Enable MTP layer parameter updates during training. | `False` | bool flag (set to enable) | Miles Native |
| `--mtp-num-layers` | Number of MTP layers to include. | `None` | Type: int | Megatron-LM (Reset by Miles) |
| `--mtp-loss-scaling-factor` | Scaling factor applied to the MTP loss. | `0.2` | Type: float | Megatron-LM (Reset by Miles) |

---

## SGLang Backend Arguments

Most SGLang server arguments can be passed through by adding the `--sglang-` prefix (some are intentionally skipped, e.g. `model_path`, `tp_size`, `port`, `nnodes`, `node_rank`). For a full list, refer to the [SGLang Server Arguments documentation](https://docs.sglang.io/advanced_features/server_arguments.html).

Commonly used arguments:

| Argument | Description | Default | Options | Source |
| :--- | :--- | :--- | :--- | :--- |
| `--sglang-mem-fraction-static` | Fraction of GPU memory to reserve for SGLang KV cache. | `0.9` | Type: float | SGLang |
| `--sglang-server-concurrency` | Maximum number of concurrent requests. | `512` | Type: int | SGLang |
| `--sglang-router-ip` | IP address of the SGLang router. | `None` | Type: str | SGLang Gateway |

> | `--sglang-router-port` | Port of the SGLang router. | `None` | Type: int | SGLang Gateway |

Reviewer: And miles router.

| `--lr-warmup-iters` | Number of iterations for warmup. | `0` | Type: int | Megatron-LM |
| `--lr-decay-iters` | Number of iterations for learning rate decay. | `None` | Type: int | Megatron-LM |
| `--lr-warmup-fraction` | Fraction of total steps to warmup. | `None` | Type: float | Megatron-LM |
| `--adam-beta1` | Beta1 for Adam optimizer. | `0.9` | Type: float | Megatron-LM |
| `--adam-beta2` | Beta2 for Adam optimizer. | `0.95` | Type: float | Megatron-LM |
| `--adam-eps` | Epsilon for Adam optimizer. | `1e-8` | Type: float | Megatron-LM |
| `--true-on-policy-mode` | Strictly align SGLang's log probs and training engine's log probs to bit-wise equal. This parameter is only used for FSDP right now. [Ref](https://github.com/zhaochenyang20/Awesome-ML-SYS-Tutorial/blob/main/rlhf/slime/mismatch/blog-en.md#truly-on-policy-training) | `False` | bool flag (set to enable) | Miles Native |
| `--train-env-vars` | Extra environment variables for training process, e.g., PyTorch memory management ones. | `{}` | Type: JSON / Dict | Miles Native |
| `--train-memory-margin-bytes` | Reserved memory margin for training in bytes. Defaults to 1GB. | `1073741824` | Type: int | Miles Native |

> | `--disable-weights-backuper` | Disables the system that backups model weights (Actor, Ref, Old Actor) to CPU RAM. Disabling saves significant host memory but prevents weight-swapping features like KL-divergence. | `False` | bool flag (set to disable) | Miles Native |

Reviewer 2: ```suggestion
| `--disable-weights-backuper` | Applies to `megatron` training backend only. Disables the system that backups model weights (Actor, Ref, Old Actor) to CPU RAM. Disabling saves significant host memory but prevents features that rely on weight-swapping, such as computing KL-divergence against a reference model. **Note**: do not set `--ref-load` and `--keep-old-actor` if disable weights backuper. | `False` | bool flag (set to disable) | Miles Native |
```

| `--custom-model-provider-path` | Path to a custom function that replaces the default model provider. [Detail](../get_started/customization.md#20-model-provider---custom-model-provider-path) | `None` | Type: str | Miles Native |
| `--recompute-loss-function` | Enable recomputing the loss function to save memory during training. | `False` | bool flag (set to enable) | Miles Native |
| `--log-probs-chunk-size` | Specifies the chunk size for logprobs computation to reduce peak memory usage. Processing logits in smaller batches, it prevents CUDA OOM errors during long-context prefilling or re-computation. Set to `-1` to disable chunking. [Ref](https://github.com/sgl-project/sglang/pull/6318) | `-1` | Type: int | Miles Native |

> | `--keep-old-actor` | Maintains a "Model Queue" (Actor, Rollout Actor, Old Actor) to ensure importance sampling ratios are calculated against the exact policy version that generated the data. Essential for asynchronous RL or high-throughput scenarios where training and inference versions may diverge. **Trade-off:** Consumes additional Host Memory (~1x model size in System RAM). | `False` | bool flag (set to enable) | Miles Native |

Reviewer 2: I think the description here is not accurate. If the `update_weights_interval == 1`, it will also save a copy of the `rollout_actor`. I suggest the following changes:
```suggestion
| `--keep-old-actor` | Maintains a "Model Queue" (Actor, Rollout Actor, Old Actor) to ensure importance sampling ratios are calculated against the exact policy version that generated the data. Essential for asynchronous RL where training and inference are decoupled, preventing mathematical incorrectness due to model staleness. It consumes additional Host Memory (extra ~1x model size for `update_weights_interval > 1` or 2x for `update_weights_interval == 1`) depending on update interval. | Miles Native |
```

| `--rollout-sample-filter-path` | Path to a function that marks individual samples to be excluded from loss calculation. See [customization](../get_started/customization.md#6-rollout-sample-filter---rollout-sample-filter-path) for more details. | `None` | Type: str | Miles Native |
| `--rollout-all-samples-process-path` | Path to a function to process all samples (including filtered ones) after rollout. See [customization](../get_started/customization.md#7-rollout-all-samples-process---rollout-all-samples-process-path) for more details. | `None` | Type: str | Miles Native |

---

## Data Arguments

Arguments for dataset configuration, prompt mapping, and training batch sizes.

| Argument | Description | Default | Options | Source |
| :--- | :--- | :--- | :--- | :--- |
| `--prompt-data` | Path to the prompt dataset (JSONL format) and each line should contains `--input-key` and `--label-key` which will be used as the prompt and the label respectively. If you want to use a custom template, you can set `--apply-chat-template` to true | `None` | Type: str | Miles Native |

> | `--disable-rollout-global-dataset` | Disable the global dataset for rollout. If set, the rollout will use the `--prompt-data` as the prompt dataset, and the prompts for rollout will be sampled from the dataset. If not set, you need to manage the data by your self. | `False` | bool flag (set to disable) | Miles Native |

Reviewer 2: ```suggestion
| `--disable-rollout-global-dataset` | Disable the global dataset for rollout. By default, Miles loads `--prompt-data` into a global dataset and samples from it for rollout. Setting this flag turns off this behavior, Use this flag only when providing a custom `--rollout-function-path` (and usually a custom `--data-source-path`) that handles data loading independently. | `False` | bool flag (set to disable) | Miles Native |
```

| `--max-tokens-per-gpu` | The maximum number of tokens (Prompt + Response combined) per GPU for dynamic batch size. This parameter defines the total sequence length budget for packing samples into micro-batches during training. Note that when enabling context parallel (CP), the effective capacity is shared, so the value should be approximately `(Total_Sequence_Length) // cp_size`. | `None` | Type: int | Miles Native |
| `--log-probs-max-tokens-per-gpu` | The maximum number of tokens per GPU for calculating log probs. This is used to calculate the log probs of the responses during rollout, and should be set to a larger value than `max_tokens_per_gpu` if you want better performance. | `None` | Type: int | Miles Native |

> | `--balance-data` | Balance the number of tokens between data parallel ranks with `karmarkar_karp` for verl. Note that this may allocate the different response of the same prompt into different training steps. | `False` | Type: bool | Megatron-LM |

Reviewer 2: ```suggestion
| `--balance-data` | Repartition each rollout batch so each data-parallel rank gets a similar total token count via Karmarkar-Karp method. It may be beneficial for training speed but changes per-rank sample grouping and adds a small CPU scheduling overhead. | `False` | Type: bool | Miles Native |
```

Reviewer 2: Hi @Ratish1 , can you also change the help in arguments.py accordingly?

| `--eval-max-prompt-len` | Maximum prompt length for evaluation. | `None` | Type: int | Miles Native |
| `--eval-min-new-tokens` | Minimum tokens to generate for evaluation responses (Not used). | `None` | Type: int | Miles Native |
| `--eval-max-context-len` | Maximum context length for evaluation (defaults to rollout max context length if not set). | `None` | Type: int | Miles Native |
| `--eval-function-path` | Path to a custom evaluation function. See [customization](../get_started/customization.md#16-evaluation-function---eval-function-path) for more details. | `None` | Type: str | Miles Native |
| `--eval-input-key` | JSON key for input text in evaluation datasets. | `None` | Type: str | Miles Native |
| `--eval-label-key` | JSON key for ground truth labels in evaluation datasets. | `None` | Type: str | Miles Native |
| `--eval-tool-key` | JSON key for tool definitions in evaluation datasets. | `None` | Type: str | Miles Native |

---

## Checkpointing and Resuming

Arguments for saving and loading model states.

| Argument | Description | Default | Options | Source |
| :--- | :--- | :--- | :--- | :--- |

> | `--load` | Path to the training model checkpoint to load. | `None` | Type: str | Megatron-LM |

Author: same

| `--save` | Path to save checkpoints. | `None` | Type: str | Megatron-LM |
| `--save-interval` | Interval (in rollout steps) to save checkpoints. Requires `--save` to be set. | `None` | Type: int | Megatron-LM |
| `--async-save` | Enable asynchronous checkpoint saving (Megatron backend only). | `False` | bool flag (set to enable) | Megatron-LM |
| `--save-hf` | Path to save the model in HuggingFace format when using Megatron backend. The model will be saved to `save_hf.format(rollout_id)`. | `None` | Type: str | Miles Native |
| `--no-save-optim` | If set, optimizer state is not saved with checkpoints to reduce size, but prevents resumption of training. | `False` | bool flag (set to enable) | Miles Native |
| `--ref-load` | Path to the reference model checkpoint. Used as initial checkpoint if `--load` is not set. | `None` | Type: str | Miles Native |
| `--ref-ckpt-step` | The checkpoint step for reference model. | `None` | Type: int | Miles Native |
| `--critic-load` | Checkpoint to load for the critic model. | value of `--load` | Type: str | Miles Native |
| `--critic-save` | Path to save the critic model. | `None` | Type: str | Miles Native |

> | `--start-rollout-id` | The starting rollout step, if not set, will try to load the step from `--load` when doing continue training, otherwise will be set to `0`, meaning training from start. | `None` | Type: int | Miles Native |

Author: done

| `--critic-lr` | Learning rate for the Critic. Defaults to `--lr`. | `None` | Type: float | Miles Native |
| `--critic-lr-warmup-iters` | Number of iterations for Critic learning rate linear warmup. | `0` | Type: int | Miles Native |
| `--num-critic-only-steps` | Number of initial steps dedicated to training only the Critic. | `0` | Type: int | Miles Native |
| `--eps-clip` | PPO clip range. | `0.2` | Type: float | Miles Native |
| `--eps-clip-high` | PPO clip upper range (defaults to `--eps-clip` if not set). | `None` | Type: float | Miles Native |
| `--eps-clip-c` | Lower bound for [Dual-clip PPO](https://arxiv.org/pdf/1912.09729). | `None` | Type: float | Miles Native |
| `--value-clip` | Clip range for value loss. | `0.2` | Type: float | Miles Native |
| `--kl-coef` | KL penalty coefficient for reward shaping. This is applied to the reward signal before advantage calculation. | `0.00` | Type: float | Miles Native |
| `--use-kl-loss` | Enable KL loss term in the final objective (as in GRPO). | `False` | bool flag (set to enable) | Miles Native |
| `--kl-loss-coef` | Weight of the KL loss term in the final objective. | `0.0` | Type: float | Miles Native |

> | `--kl-loss-type` | Selection of the KL loss implementation. | `k1` | `k1`, `k2`, `k3`, `low_var_kl` | Miles Native |

Reviewer 2: ```suggestion
| `--kl-loss-type` | Selection of the KL loss implementation. `k1`: standard approximation, `k2`: quadratic approximation, `k3`: [Monte-Carlo approximation](http://joschu.net/blog/kl-approx.html), `low_var_kl`: Same logic as `k3`, but explicitly clamped to prevent numerical explosions. | `k1` | `k1`, `k2`, `k3`, `low_var_kl` | Miles Native |
```

| `--use-unbiased-kl` | Enable unbiased KL estimation. | `False` | bool flag (set to enable) | Miles Native |
| `--entropy-coef` | Coefficient for entropy regularization. | `0.0` | Type: float | Miles Native |

> | `--gamma` | PPO GAE gamma. | `1.0` | Type: float | Miles Native |

Reviewer 2: ```suggestion
| `--gamma` | Discount factor for future rewards. Used in PPO (GAE) and REINFORCE++. | `1.0` | Type: float | Miles Native |
```

| `--tis-clip` | Clipping threshold C for importance sampling ratios to control variance. | `2.0` | Type: float | Miles Native |
| `--tis-clip-low` | Lower bound clipping threshold C for importance sampling ratios to control variance. | `0.0` | Type: float | Miles Native |
| `--custom-tis-function-path` | Path to a custom TIS or MIS function. See [customization](../get_started/customization.md#10-custom-tisrs-function---custom-tis-function-path) for more details. | `None` | Type: str | Miles Native |
| `--custom-pg-loss-reducer-function-path` | Custom reducer function for policy gradient loss. See [customization](../get_started/customization.md#11-custom-pg-loss-reducer---custom-pg-loss-reducer-function-path) for more details. | `None` | Type: str | Miles Native |
| `--use-routing-replay` | Enable [Routing Replay](https://arxiv.org/abs/2507.18071). | `False` | bool flag (set to enable) | Miles Native |
| `--use-rollout-routing-replay` | Enable R3: [Rollout Routing Replay](https://arxiv.org/pdf/2510.11370). | `False` | bool flag (set to enable) | Miles Native |

> | `--use-opsm` | Enable Off-Policy Sequence Masking (OPSM). | `False` | bool flag (set to enable) | Miles Native |

Reviewer 2: ```suggestion
| `--use-opsm` | Enable Off-Policy Sequence Masking (OPSM). Filters sequences that have **BOTH** negative advantages (bad results) AND high KL divergence (stale data). This stabilizes training by preventing updates from unreliable, highly off-policy samples. | `False` | bool flag (set to enable) | Miles Native |
```

| `--opsm-delta` | The threshold for Off-Policy Sequence Masking (OPSM). | `1e-4` | Type: float | Miles Native |
| `--get-mismatch-metrics` | Whether to calculate the mismatch metrics. It will **only return mismatch metrics** but not change the loss in any way. | `False` | bool flag (set to enable) | Miles Native |
| `--ref-update-interval` | Interval (in rollout steps) to update ref model from actor. If `None`, ref model is not updated. | `None` | Type: int | Miles Native |
| `--reset-optimizer-states` | Resets the optimizer state after each rollout round. This clears the optimization history, which can improve stability or satisfy specific experimental requirements. | `False` | bool flag (set to enable) | Miles Native |
| `--disable-grpo-std-normalization` | Disable standard deviation normalization for GRPO. From [Dr.GRPO](https://arxiv.org/pdf/2503.20783) | `False` | bool flag (set to enable) | Miles Native |
| `--disable-rewards-normalization` | Disable rewards normalization. | `False` | bool flag (set to enable) | Miles Native |

> | `--use-rollout-entropy` | Enable entropy calculation when calculating the logprobs from actor and reference model. This is useful for doing special loss mask. | `False` | bool flag (set to enable) | Miles Native |

Reviewer 2: ```suggestion
| `--use-rollout-entropy` | Enable entropy calculation when calculating the logprobs from actor and reference model. This is useful for implementing custom entropy-based loss masking. | `False` | bool flag (set to enable) | Miles Native |
```

> | `--use-rollout-logprobs` | Use rollout logprobs for importance sampling ratios, use the logprobs from the actor model if not set. | `False` | bool flag (set to enable) | Miles Native |

Reviewer 2: ```suggestion
| `--use-rollout-logprobs` | Use rollout logprobs for importance sampling ratios, use the logprobs from the actor model if not set. If `--get-mismatch-metrics` is set, the log probs will be recomputed by training engine, one more forward pass will be applied. | `False` | bool flag (set to enable) | Miles Native |
```
